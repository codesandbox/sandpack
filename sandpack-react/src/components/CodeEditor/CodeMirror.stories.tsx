import * as React from "react";

import { SandpackProvider } from "../../contexts/sandpackContext";
import { SandpackThemeProvider } from "../../contexts/themeContext";

import { CodeEditor } from "./index";

export default {
  title: "components/CodeMirror",
  component: CodeEditor,
};

export const JustEditor: React.FC = () => (
  <SandpackProvider template="vue">
    <SandpackThemeProvider>
      <CodeEditor
        code="const c = a+b;"
        fileType="jsx"
        onCodeUpdate={(): void => console.log("code update")}
      />
    </SandpackThemeProvider>
  </SandpackProvider>
);
