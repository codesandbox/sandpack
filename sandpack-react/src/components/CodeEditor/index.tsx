import { useClasser } from "@code-hike/classer";
import type { Extension } from "@codemirror/state";
import type { KeyBinding } from "@codemirror/view";
import * as React from "react";

import { RunButton } from "../../common/RunButton";
import { SandpackStack } from "../../common/Stack";
import { useActiveCode } from "../../hooks/useActiveCode";
import { useSandpack } from "../../hooks/useSandpack";
import type { SandpackInitMode } from "../../types";
import { FileTabs } from "../FileTabs";

import { CodeMirror } from "./CodeMirror";
import type { CodeMirrorRef } from "./CodeMirror";

export type CodeEditorRef = CodeMirrorRef;
export interface CodeEditorProps {
  customStyle?: React.CSSProperties;
  showTabs?: boolean;
  showLineNumbers?: boolean;
  showInlineErrors?: boolean;
  showRunButton?: boolean;
  wrapContent?: boolean;
  closableTabs?: boolean;
  /**
   * This provides a way to control how some components are going to
   * be initialized on the page. The CodeEditor and the Preview components
   * are quite expensive and might overload the memory usage, so this gives
   * a certain control of when to initialize them.
   */
  initMode?: SandpackInitMode;
  /**
   * CodeMirror extensions for the editor state, which can
   * provide extra features and functionalities to the editor component.
   */
  extensions?: Extension[];
  /**
   * Property to register CodeMirror extension keymap.
   */
  extensionsKeymap?: Array<readonly KeyBinding[]>;
  id?: string;
  /**
   * This disables editing of the editor content by the user.
   */
  readOnly?: boolean;
  /**
   * Controls the visibility of Read-only label, which will only
   * appears when `readOnly` is `true`
   */
  showReadOnly?: boolean;
}

export { CodeMirror as CodeEditor };

/**
 * @category Components
 */
export const SandpackCodeEditor = React.forwardRef<
  CodeMirrorRef,
  CodeEditorProps
>(
  (
    {
      customStyle,
      showTabs,
      showLineNumbers = false,
      showInlineErrors = false,
      showRunButton = true,
      wrapContent = false,
      closableTabs = false,
      initMode,
      extensions,
      extensionsKeymap,
      id,
      readOnly,
      showReadOnly,
    },
    ref
  ) => {
    const { sandpack } = useSandpack();
    const { code, updateCode, readOnly: readOnlyFile } = useActiveCode();
    const { activePath, status, editorState } = sandpack;
    const shouldShowTabs = showTabs ?? sandpack.openPaths.length > 1;

    const c = useClasser("sp");

    const handleCodeUpdate = (newCode: string): void => {
      updateCode(newCode);
    };

    return (
      <SandpackStack customStyle={customStyle}>
        {shouldShowTabs && <FileTabs closableTabs={closableTabs} />}

        <div className={c("code-editor")}>
          <CodeMirror
            key={activePath}
            ref={ref}
            code={code}
            editorState={editorState}
            extensions={extensions}
            extensionsKeymap={extensionsKeymap}
            filePath={activePath}
            id={id}
            initMode={initMode || sandpack.initMode}
            onCodeUpdate={handleCodeUpdate}
            readOnly={readOnly || readOnlyFile}
            showInlineErrors={showInlineErrors}
            showLineNumbers={showLineNumbers}
            showReadOnly={showReadOnly}
            wrapContent={wrapContent}
          />

          {showRunButton && status === "idle" ? <RunButton /> : null}
        </div>
      </SandpackStack>
    );
  }
);
