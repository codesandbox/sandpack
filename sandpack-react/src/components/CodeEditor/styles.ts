import { css } from "../../styles";
import { buttonClassName } from "../../styles/shared";

import { styleTokens } from "./utils";

export const placeholderClassName = css({
  margin: "0",
  display: "block",
  fontFamily: "$font$mono",
  fontSize: "$font$size",
  color: "$syntax$color$plain",
  lineHeight: "$font$lineHeight",
});

export const tokensClassName = css(styleTokens());

export const editorClassName = css({
  flex: 1,
  position: "relative",
  overflow: "auto",
  background: "$colors$surface1",

  ".cm-scroller": {
    padding: "$space$4 0",
  },

  [`.${placeholderClassName}`]: {
    padding: "$space$4 0",
  },

  /**
   * For iOS: prevent browser zoom when clicking on sandbox.
   * Does NOT apply to code blocks.
   */
  "@media screen and (max-width: 768px)": {
    "@supports (-webkit-overflow-scrolling: touch)": {
      ".cm-content": { fontSize: "16px" },
    },
  },
});

export const cmClassName = css({
  margin: "0",
  outline: "none",
  height: "100%",
});

export const readOnlyClassName = css({
  fontFamily: "$font$mono",
  fontSize: "0.8em",
  position: "absolute",
  right: "$space$2",
  bottom: "$space$2",
  zIndex: "$top",
  color: "$colors$clickable",
  backgroundColor: "$colors$surface2",
  borderRadius: "99999px",
  padding: "calc($space$1 / 2) $space$2",

  [`& + .${buttonClassName}`]: {
    right: "calc($space$11 * 2)",
  },
});
