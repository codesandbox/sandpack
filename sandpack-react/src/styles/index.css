.sp-wrapper {
  --sp-colors-fg-active: #1f2933;
  --sp-colors-fg-default: #757678;
  --sp-colors-fg-inactive: #e4e7eb;
  --sp-colors-bg-active: #e4e7eb;
  --sp-colors-bg-default: #f8f9fb;
  --sp-colors-bg-default-overlay: #f8f9fbcf;
  --sp-colors-bg-input: #ffffff;
  --sp-colors-accent: #64d2ff;
  --sp-colors-bg-error: #ffcdca;
  --sp-colors-fg-error: #811e18;
  --sp-layout-height: 300px;

  --sp-font-size: 14px;
  --sp-font-body: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto,
    Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji",
    "Segoe UI Symbol";
  --sp-font-mono: "Fira Mono", "DejaVu Sans Mono", Menlo, Consolas,
    "Liberation Mono", Monaco, "Lucida Console", monospace;

  --sp-space-1: 4px;
  --sp-space-2: 8px;
  --sp-space-3: 12px;
  --sp-space-4: 16px;
  --sp-space-5: 20px;
  --sp-space-6: 24px;
  --sp-space-7: 28px;
  --sp-space-8: 32px;
  --sp-border-radius: 4px;

  all: initial;

  font-size: var(--sp-font-size);
  font-family: var(--sp-font-body);
  display: block;
  box-sizing: border-box;

  text-rendering: optimizeLegibility;
  -webkit-tap-highlight-color: transparent;
  -webkit-font-smoothing: subpixel-antialiased;
}

@media screen and (min-resolution: 2dppx) {
  .sp-wrapper {
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }
}

.sp-wrapper * {
  box-sizing: border-box;
}

.sp-wrapper:focus {
  outline: 0;
}

.sp-wrapper *::-webkit-scrollbar {
  width: 8px;
  height: 8px;
}

.sp-wrapper *::-webkit-scrollbar-track {
  background-color: var(--sp-colors-bg-default);
}

.sp-wrapper *::-webkit-scrollbar-corner {
  background-color: transparent;
}

.sp-wrapper *::-webkit-scrollbar-thumb {
  background-color: var(--sp-colors-fg-default);
  border-radius: 9999px;
  opacity: 0;
}

.sp-wrapper *::-webkit-scrollbar-thumb:hover {
  opacity: 1;
}

.sp-wrapper *::-webkit-scrollbar-track {
  border-left: 1px solid var(--sp-colors-fg-inactive);
}

.sp-layout {
  border: 1px solid var(--sp-colors-fg-inactive);
  display: flex;
  flex-wrap: wrap;
  align-items: stretch;
  background: var(--sp-colors-bg-default);
  border-radius: var(--sp-border-radius);
  overflow: hidden;
  -webkit-mask-image: -webkit-radial-gradient(
    var(--sp-colors-bg-input),
    var(--sp-colors-bg-default)
  );
}

.sp-stack {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.sp-layout > .sp-stack {
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: 0;
  min-width: 350px;
  height: var(--sp-layout-height);
}

@media print {
  .sp-layout > .sp-stack {
    height: auto;
    display: block;
  }
}

.sp-layout > *:not(:first-child) {
  border-left: 1px solid var(--sp-colors-fg-inactive);
  border-top: 1px solid var(--sp-colors-fg-inactive);
  margin-left: -1px;
  margin-top: -1px;
  position: relative;
}

.sp-layout > *:first-child .sp-overlay {
  border-right: 1px solid var(--sp-colors-fg-inactive);
}

@media screen and (max-width: 768px) {
  .sp-layout > .sp-stack {
    height: auto;
    min-width: 100% !important; /* triggers the layout break at the 768px breakpoint, not when the component is less then 700px */
  }
}

/* Components styling */

.sp-pre-placeholder {
  margin: 0;
  display: block;
  padding: 0 var(--sp-space-3);
  font-family: var(--sp-font-mono);
  font-size: var(--sp-font-size);
  color: var(--sp-colors-fg-active);
  line-height: var(--sp-line-height);
}

.sp-code-editor {
  flex: 1;
  position: relative;
  overflow: auto;
  background: var(--sp-colors-bg-default);
}

.sp-cm {
  padding: var(--sp-space-4) 0;
  margin: 0;
  outline: none;
  height: 100%;
}

.sp-cm:focus-visible {
  box-shadow: inset 0 0 0 4px var(--sp-colors-accent);
  padding-left: var(--sp-space-1);
  padding-right: var(--sp-space-1);
}

.sp-cm:focus-visible .cm-line {
  padding: 0 var(--sp-space-2);
}

.sp-cm:focus-visible .cm-gutter.cm-lineNumbers {
  padding-left: 0;
  padding-right: var(--sp-space-2);
}

.sp-transpiled-code {
  display: flex;
  flex-direction: column;
  width: 100%;
  position: relative;
  overflow: auto;
  min-height: 160px;
  flex: 1;
}

.sp-code-view {
  background: var(--sp-colors-bg-default);
  padding: var(--sp-space-4) var(--sp-space-2);
  line-height: 1.4;
  overflow: auto;
  flex: 1;
  position: relative;
}

.sp-tabs {
  border-bottom: 1px solid var(--sp-colors-fg-inactive);
  background: var(--sp-colors-bg-default);
}

.sp-tabs-scrollable-container {
  padding: 0 var(--sp-space-4);
  overflow: auto;
  display: flex;
  flex-wrap: nowrap;
  align-items: stretch;
  min-height: 40px;
  margin-bottom: -1px;
}

.sp-preview-container {
  flex: 1;
  display: flex;
  flex-direction: column;
  background: white;
  overflow: auto;
  position: relative;
}

.sp-preview-iframe {
  border: 0;
  outline: 0;
  width: 100%;
  height: 100%;
  min-height: 160px; /* fits a minimal the error screen */
  max-height: 2000px; /* safety measure */
  flex: 1;
  transition: height 0.15s ease-out;
}

.sp-preview-actions {
  display: flex;
  position: absolute;
  bottom: var(--sp-space-2);
  right: var(--sp-space-2);
  z-index: 4;
}

.sp-preview-actions > * {
  margin-left: var(--sp-space-2);
}

.sp-navigator {
  display: flex;
  align-items: center;
  height: 40px;
  border-bottom: 1px solid var(--sp-colors-fg-inactive);
  padding: var(--sp-space-2) var(--sp-space-4);
  background: var(--sp-colors-bg-default);
}

/* Common Styling */

.sp-tab-button {
  display: block;
  background: transparent;
  appearance: none;
  font-size: inherit;
  padding: 0 var(--sp-space-2);
  color: var(--sp-colors-fg-default);
  height: 40px;
  border: 0;
  outline: none;
  border-bottom: 1px solid transparent;
  transition: all 0.15s ease-out;
  white-space: nowrap;
}

.sp-tab-button[data-active="true"] {
  color: var(--sp-colors-fg-active);
  border-bottom: 1px solid var(--sp-colors-accent);
}

.sp-tab-button:hover {
  color: var(--sp-colors-fg-active);
  background: var(--sp-colors-bg-active);
}

.sp-tab-button:focus {
  outline: none;
}

.sp-tab-button:focus-visible {
  box-shadow: inset 0 0 0 2px var(--sp-colors-accent);
}

.sp-close-button {
  padding: 0px var(--sp-space-1) 2px var(--sp-space-1);
  border-radius: var(--sp-border-radius);
  margin-left: var(--sp-space-1);
  width: 20px;
  visibility: hidden;
}

.sp-tab-button:hover > .sp-close-button {
  visibility: unset;
}

.sp-button {
  appearance: none;
  border: 0;
  outline: none;
  padding: var(--sp-space-1) var(--sp-space-3) var(--sp-space-1)
    var(--sp-space-2);
  border-radius: var(--sp-border-radius);
  display: flex;
  align-items: center;
  color: var(--sp-colors-fg-default);
  background-color: var(--sp-colors-bg-default);
  font-size: inherit;
  font-family: inherit;
  transition: all 0.15s ease-in-out;
}

.sp-button.sp-explorer {
  border-radius: 0;
  width: 100%;
  height: 28px;
}

.sp-button.sp-explorer svg {
  margin-right: var(--sp-space-1);
}

.sp-button.sp-icon-standalone {
  padding: var(--sp-space-1);
  background: var(--sp-colors-bg-default-overlay);
  backdrop-filter: blur(4px);
  width: var(--sp-space-8);
  height: var(--sp-space-8);
}

.sp-button.sp-icon {
  padding: 0;
}

.sp-button:hover:not(:disabled) {
  background-color: var(--sp-colors-bg-active);
  color: var(--sp-colors-fg-active);
}

.sp-button:disabled {
  color: var(--sp-colors-fg-inactive);
}

.sp-button:focus {
  outline: none;
}

.sp-button:focus-visible {
  outline: 2px solid var(--sp-colors-accent);
}

.sp-button.sp-csb-icon-dark {
  color: var(--sp-colors-bg-default);
}

.sp-button.sp-csb-icon-dark:hover:not(:disabled) {
  background-color: var(--sp-colors-bg-active);
  color: var(--sp-colors-bg-active);
}

.sp-input {
  background-color: var(--sp-colors-bg-input);
  color: var(--sp-colors-fg-default);
  padding: var(--sp-space-1) var(--sp-space-2);
  border-radius: var(--sp-border-radius);
  border: 1px solid var(--sp-colors-fg-inactive);
  height: 24px;
  line-height: 24px;
  font-size: inherit;
  outline: none;
  flex: 1;
  margin-left: var(--sp-space-4);
  width: 0;
}

.sp-input:focus {
  border: 1px solid var(--sp-colors-accent);
  color: var(--sp-colors-fg-active);
}

.sp-overlay {
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  margin: 0;
  overflow: auto;
  height: 100%;
  z-index: 3;
}

.sp-loading {
  background-color: var(--sp-colors-bg-default);
  z-index: 5;
}

.sp-cube-wrapper {
  position: absolute;
  right: var(--sp-space-2);
  bottom: var(--sp-space-2);
  z-index: 4;
  width: 32px;
  height: 32px;
  background: var(--sp-colors-bg-default);
  border-radius: var(--sp-border-radius);
}

.sp-cube-wrapper .sp-cube {
  display: block;
}

.sp-cube-wrapper .sp-button {
  display: none;
}

.sp-cube-wrapper:hover .sp-button {
  display: block;
}

.sp-cube-wrapper:hover .sp-cube {
  display: none;
}

.sp-cube {
  transform: translate(-4px, 9px) scale(0.13, 0.13);
}

.sp-cube * {
  position: absolute;
  width: 96px;
  height: 96px;
}
.sp-cube .sp-sides * {
  border: 10px solid var(--sp-colors-fg-default);
  border-radius: 8px;
  background: var(--sp-colors-bg-default-overlay);
}

@keyframes sp-cube-rotate {
  0% {
    transform: rotateX(-25.5deg) rotateY(45deg);
  }

  100% {
    transform: rotateX(-25.5deg) rotateY(405deg);
  }
}

.sp-cube .sp-sides {
  animation: sp-cube-rotate 1s linear infinite;
  animation-fill-mode: forwards;
  transform-style: preserve-3d;
  transform: rotateX(-25.5deg) rotateY(45deg);
}

.sp-cube .sp-sides .sp-top {
  transform: rotateX(90deg) translateZ(44px);
  transform-origin: 50% 50%;
}

.sp-cube .sp-sides .sp-bottom {
  transform: rotateX(-90deg) translateZ(44px);
  transform-origin: 50% 50%;
}

.sp-cube .sp-sides .sp-front {
  transform: rotateY(0deg) translateZ(44px);
  transform-origin: 50% 50%;
}

.sp-cube .sp-sides .sp-back {
  transform: rotateY(-180deg) translateZ(44px);
  transform-origin: 50% 50%;
}

.sp-cube .sp-sides .sp-left {
  transform: rotateY(-90deg) translateZ(44px);
  transform-origin: 50% 50%;
}

.sp-cube .sp-sides .sp-right {
  transform: rotateY(90deg) translateZ(44px);
  transform-origin: 50% 50%;
}

@keyframes sp-fade-in {
  0% {
    opacity: 0;
    transform: translateY(4px);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

.sp-error {
  padding: var(--sp-space-4);
  white-space: pre-wrap;
  font-family: var(--sp-font-mono);
  background-color: var(--sp-colors-bg-error);
}

.sp-error-message {
  animation: sp-fade-in 0.15s ease-in;
  color: var(--sp-colors-fg-error);
}

.sp-devtools {
  height: var(--sp-layout-height);
  width: 100%;
}

.sp-read-only {
  font-family: var(--sp-font-mono);
  font-size: 0.8em;
  position: absolute;
  right: var(--sp-space-2);
  bottom: var(--sp-space-2);
  z-index: 2;
  color: var(--sp-colors-bg-active);
  background-color: var(--sp-colors-fg-active);
  border-radius: 99999px;
  padding: calc(var(--sp-space-1) / 2) var(--sp-space-2);
}
