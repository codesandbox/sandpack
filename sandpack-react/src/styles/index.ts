import { createStitches } from "@stitches/core";

import { defaultLight, defaultDark, SANDPACK_THEMES } from "../themes";
import type { SandpackTheme, SandpackThemeProp } from "../types";
import { isDarkColor } from "../utils/stringUtils";

import { createStitchesMock } from "./stitches-mock";

/**
 * @category Theme
 */
export const THEME_PREFIX = "sp";

const getNodeProcess = (): false | string | undefined => {
  if (typeof process !== "undefined") {
    return process.env.SANDPACK_BARE_COMPONENTS;
  }

  return false;
};

/**
 * @category Theme
 */
export const { createTheme, css, getCssText, keyframes } = getNodeProcess()
  ? createStitchesMock
  : createStitches({
      prefix: THEME_PREFIX,
    });

const defaultVariables = {
  space: new Array(11).fill(" ").reduce((acc, _, index) => {
    return { ...acc, [index + 1]: `${(index + 1) * 4}px` };
  }, {}),
  border: { radius: "4px" },
  layout: { height: "300px", headerHeight: "40px" },
  transitions: { default: "150ms ease" },
  zIndices: {
    base: "1",
    overlay: "2",
    top: "3",
  },
};

/**
 * @category Theme
 */
export const standardizeStitchesTheme = (
  theme: SandpackTheme
): Record<string, Record<string, string>> => {
  // Flat values
  const syntaxEntries = Object.entries(theme.syntax);
  const syntax = syntaxEntries.reduce((tokenAcc, [tokenName, tokenValue]) => {
    // Single property
    let newValues = { [`color-${tokenName}`]: tokenValue };

    // Multiples properties
    if (typeof tokenValue === "object") {
      newValues = Object.entries(tokenValue).reduce(
        (valueAcc, [styleProp, styleValue]) => {
          return {
            ...valueAcc,
            [`${styleProp}-${tokenName}`]: styleValue,
          };
        },
        {}
      );
    }

    return { ...tokenAcc, ...newValues };
  }, {});

  return {
    ...defaultVariables,
    colors: theme.colors,
    font: theme.font,
    syntax,
  };
};

/**
 * @category Theme
 */
export const standardizeTheme = (
  inputTheme: SandpackThemeProp = "light"
): { id: string; theme: SandpackTheme; mode: "dark" | "light" } => {
  const defaultLightThemeKey = "default";

  /**
   * Set a local theme: dark or light
   */
  if (typeof inputTheme === "string") {
    const predefinedTheme = SANDPACK_THEMES[inputTheme];
    if (!predefinedTheme) {
      throw new Error(
        `[sandpack-react]: invalid theme '${inputTheme}' provided.`
      );
    }

    return {
      theme: predefinedTheme,
      id: inputTheme,
      mode: "light",
    };
  }

  /**
   * Fullfill the colors key, in case it's missing any key
   */
  const mode = isDarkColor(
    inputTheme?.colors?.surface1 ?? defaultLight.colors.surface1
  )
    ? "dark"
    : "light";

  /**
   * Figure out what's the properly default colors it should be
   * error, warning and success colors have different values between dark and light
   */
  const baseTheme = mode === "dark" ? defaultDark : defaultLight;
  const colorsByMode = { ...baseTheme.colors, ...(inputTheme?.colors ?? {}) };
  const syntaxByMode = { ...baseTheme.syntax, ...(inputTheme?.syntax ?? {}) };
  const fontByMode = { ...baseTheme.font, ...(inputTheme?.font ?? {}) };

  const theme = {
    colors: colorsByMode,
    syntax: syntaxByMode,
    font: fontByMode,
  };

  const id = inputTheme
    ? simpleHashFunction(JSON.stringify(theme))
    : defaultLightThemeKey;

  return {
    theme,
    id: `sp-${id}`,
    mode,
  };
};

const simpleHashFunction = (str: string): number => {
  let hash = 0;
  for (let i = 0; i < str.length; hash &= hash) {
    hash = 31 * hash + str.charCodeAt(i++);
  }
  return Math.abs(hash);
};
