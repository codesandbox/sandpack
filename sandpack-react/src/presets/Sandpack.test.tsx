/* eslint-disable @typescript-eslint/no-var-requires */
import React from "react";
import { renderToString } from "react-dom/server";

import { Sandpack } from "../";
import { editorClassName } from "../components/CodeEditor/styles";

describe("Sandpack", () => {
  it("renders on a server without crashing", () => {
    const renderOnServer = (): string =>
      renderToString(<Sandpack options={{ id: "test" }} />);
    expect(renderOnServer).not.toThrow();
  });
});

describe("getSandpackCssText", () => {
  it("should return only the essential style", () => {
    let sandpackModule = null;
    let renderToString = null;

    const componentClassName = editorClassName().className;

    jest.isolateModules(() => {
      sandpackModule = require("../");
      renderToString = require("react-dom/server").renderToString;
    });

    const { getSandpackCssText, SandpackProvider } = sandpackModule;

    renderToString(
      <SandpackProvider>
        <div />
      </SandpackProvider>
    );

    expect(getSandpackCssText().length).toBe(4554);
    expect(getSandpackCssText()).not.toContain(componentClassName);
  });

  it("should return the essential style + the component style", () => {
    let sandpackModule = null;
    let renderToString = null;

    jest.isolateModules(() => {
      sandpackModule = require("../");
      renderToString = require("react-dom/server").renderToString;
    });

    const { SandpackProvider, SandpackCodeEditor, getSandpackCssText } =
      sandpackModule;

    const componentClassName = editorClassName().className;

    renderToString(
      <SandpackProvider>
        <SandpackCodeEditor />
      </SandpackProvider>
    );

    expect(getSandpackCssText()).toContain(componentClassName);
  });

  it("should not duplicate the component CSS if there'is more than one component", () => {
    let sandpackModule = null;
    let renderToString = null;

    jest.isolateModules(() => {
      sandpackModule = require("../");
      renderToString = require("react-dom/server").renderToString;
    });

    const componentClassName = editorClassName().className;
    const { SandpackProvider, SandpackCodeEditor, getSandpackCssText } =
      sandpackModule;

    renderToString(
      <>
        <SandpackProvider>
          <SandpackCodeEditor />
          <SandpackCodeEditor />
        </SandpackProvider>

        <SandpackProvider>
          <SandpackCodeEditor />
        </SandpackProvider>
      </>
    );

    const selector = new RegExp(`${componentClassName}{`, "gm");
    expect(getSandpackCssText().match(selector).length).toBe(1);
  });

  it("should not duplicate the CSS if two Sandpack has the same theme", () => {
    let sandpackModule = null;
    let renderToString = null;

    jest.isolateModules(() => {
      sandpackModule = require("../");
      renderToString = require("react-dom/server").renderToString;
    });

    const { Sandpack, getSandpackCssText } = sandpackModule;

    renderToString(
      <>
        <Sandpack theme="light" />
        <Sandpack theme="light" />
      </>
    );

    expect(getSandpackCssText()).toContain("--sxs{--sxs:0 light}");
    expect(getSandpackCssText().match(/--sp-colors-surface1:/g).length).toBe(1);
  });

  it("should only include the new theme variables if a theme is different", () => {
    let sandpackModule = null;
    let renderToString = null;

    jest.isolateModules(() => {
      sandpackModule = require("../");
      renderToString = require("react-dom/server").renderToString;
    });

    const { Sandpack, getSandpackCssText } = sandpackModule;

    renderToString(
      <>
        <Sandpack theme="light" />
        <Sandpack theme="dark" />
      </>
    );

    expect(getSandpackCssText()).toContain("--sxs{--sxs:0 light dark}");
    expect(getSandpackCssText().match(/--sp-colors-surface1:/g).length).toBe(2);
  });
});
