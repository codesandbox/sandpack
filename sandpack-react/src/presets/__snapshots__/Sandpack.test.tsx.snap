// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sandpack match the snapshot on server render 1`] = `
"<div class=\\"sp-wrapper sp-default sp-c-fVPbOs\\"><div class=\\"sp-layout sp-c-dvoSuG\\"><div class=\\"sp-stack sp-c-jKpMgH\\" style=\\"flex-grow:50;flex-shrink:50;min-width:350px\\"><div class=\\"sp-tabs sp-c-dyHYiL\\" translate=\\"no\\"><div aria-label=\\"Select active file\\" class=\\"sp-tabs-scrollable-container sp-c-gGbYbQ\\" role=\\"tablist\\"><button aria-selected=\\"true\\" class=\\"sp-tab-button sp-c-jCyRbv sp-c-cFDjHt\\" data-active=\\"true\\" role=\\"tab\\" title=\\"/src/index.js\\" type=\\"button\\">index.js</button><button aria-selected=\\"false\\" class=\\"sp-tab-button sp-c-jCyRbv sp-c-cFDjHt\\" data-active=\\"false\\" role=\\"tab\\" title=\\"/src/styles.css\\" type=\\"button\\">styles.css</button><button aria-selected=\\"false\\" class=\\"sp-tab-button sp-c-jCyRbv sp-c-cFDjHt\\" data-active=\\"false\\" role=\\"tab\\" title=\\"/index.html\\" type=\\"button\\">index.html</button></div></div><div class=\\"sp-code-editor sp-c-fdnnct\\"><div aria-describedby=\\"enter-instructions-test\\" aria-label=\\"Code Editor for index.js\\" class=\\"sp-cm sp-pristine sp-javascript sp-c-jcgexo sp-c-jkvvao\\" role=\\"group\\" tabindex=\\"0\\" translate=\\"no\\"><pre class=\\"sp-pre-placeholder sp-c-cdSpLI\\" style=\\"margin-left:20px\\"><span class=\\"sp-syntax-keyword\\">import</span> <span class=\\"sp-syntax-string\\">&quot;./styles.css&quot;</span><span class=\\"sp-syntax-punctuation\\">;</span>

<span class=\\"sp-syntax-plain\\">document</span>.<span class=\\"sp-syntax-property\\">getElementById</span><span class=\\"sp-syntax-punctuation\\">(</span><span class=\\"sp-syntax-string\\">&quot;app&quot;</span><span class=\\"sp-syntax-punctuation\\">)</span>.<span class=\\"sp-syntax-property\\">innerHTML</span> = <span class=\\"sp-syntax-string\\">\`
&lt;h1&gt;Hello Vanilla!&lt;/h1&gt;
&lt;div&gt;
  We use the same configuration as Parcel to bundle this sandbox, you can find more
  info about Parcel 
  &lt;a href=&quot;https://parceljs.org&quot; target=&quot;_blank&quot; rel=&quot;noopener noreferrer&quot;&gt;here&lt;/a&gt;.
&lt;/div&gt;
\`</span><span class=\\"sp-syntax-punctuation\\">;</span></pre><p id=\\"enter-instructions-test\\" style=\\"display:none\\">To enter the code editing mode, press Enter. To exit the edit mode, press Escape</p><p id=\\"exit-instructions-test\\" style=\\"display:none\\">You are editing the code. To exit the edit mode, press Escape</p></div></div></div><div class=\\"sp-stack sp-c-jKpMgH\\" style=\\"flex-grow:50;flex-shrink:50;min-width:350px\\"><div class=\\"sp-preview-container sp-c-bIhyHd\\"><iframe class=\\"sp-preview-iframe sp-c-fgviib\\" title=\\"Sandpack Preview\\"></iframe><div class=\\"sp-preview-actions sp-c-eKnQtg\\"><a href=\\"https://codesandbox.io/api/v1/sandboxes/define?parameters=&amp;query=file=/src/index.js%26from-sandpack=true\\" rel=\\"noreferrer noopener\\" target=\\"_blank\\" title=\\"Open in CodeSandbox\\" class=\\"sp-button sp-icon-standalone sp-csb-icon-light sp-c-jCyRbv sp-c-gMfcns sp-c-eBhISN\\"><svg height=\\"16\\" viewBox=\\"0 0 16 16\\" width=\\"16\\" xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\"><path d=\\"M6.66671 3.33337H4.33337C3.78109 3.33337 3.33337 3.78109 3.33337 4.33337V11.6667C3.33337 12.219 3.78109 12.6667 4.33337 12.6667H11.6667C12.219 12.6667 12.6667 12.219 12.6667 11.6667V9.33337\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\"></path><path d=\\"M10 3.33337H12.5667C12.6219 3.33337 12.6667 3.37815 12.6667 3.43337V6.00004\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\"></path><path d=\\"M7.33337 8.66668L12.5334 3.46667\\" stroke=\\"currentColor\\" stroke-linecap=\\"round\\"></path></svg></a></div></div></div></div></div>"
`;
